### DISCUSSION TITLE
Is a super low end runner of interest?

### DISCUSSION THREAD

**Discussion by PEZ** [PlummersSoftwareLLC/Primes#799] [General]
I have an Asus Mini PC PN40, with 4GB RAM, running Windows. Windows reports the processor as being:

> Intel(R) Celeron(R) N4020 CPU @ 1.10GHz, 1101 Mhz, 2 Core(s), 2 Logical Processor(s)

Maybe the P in PN40 is short for Pentium, idk.

Anyway, the machine is super slow for the things I use it for (testing how an open source app I am building works on Windows/fixing Windows specific bugs and such. I use it very seldom and connect to it using Remote desktop. It just sits there most of the time. So I was thinking maybe it can be running sieves for this project on a regular basis?

After som fiddling around I managed to get it to run some (very few of the solutions). The output from `make` speaks mostly about  *Failing to build solution*, but some succeed. The output and the final report looks like so:

```
info: Unconfined mode: false
info: Detected architecture: amd64
warn: [Prime6502Assembly][solution_1] Skipping due to disabled build!
warn: [Prime6502Assembly][solution_2] Skipping due to disabled build!
info: [PrimeAda][solution_1] Building...
error: [PrimeAda][solution_1] Failed building solution!
info: [PrimeAmd64][solution_1] Building...
error: [PrimeAmd64][solution_1] Failed building solution!
warn: [PrimeAPL][solution_1] Skipping due to disabled build!

...

                                                                Single-threaded
┌───────┬────────────────┬──────────┬──────────────────────────────┬────────┬──────────┬─────────┬───────────┬──────────┬──────┬───────────────┐
│ Index │ Implementation │ Solution │ Label                        │ Passes │ Duration │ Threads │ Algorithm │ Faithful │ Bits │ Passes/Second │
├───────┼────────────────┼──────────┼──────────────────────────────┼────────┼──────────┼─────────┼───────────┼──────────┼──────┼───────────────┤
│   1   │ basic          │ 3        │ Nukepayload2_ReDim8of30M     │  4164  │ 5.00022  │    1    │   wheel   │   yes    │ 1    │   832.76331   │
│   2   │ csharp         │ 1        │ kinematics_rawd30            │  4083  │ 5.00110  │    1    │   wheel   │   yes    │ 1    │   816.42039   │
│   3   │ basic          │ 3        │ Nukepayload2_ArrayPool8of30M │  4051  │ 5.00038  │    1    │   wheel   │   yes    │ 1    │   810.13804   │
│   4   │ csharp         │ 1        │ kinematics_poolq30m          │  3865  │ 5.00100  │    1    │   other   │   yes    │ 1    │   772.84543   │
│   5   │ csharp         │ 1        │ kinematics_rawd6             │  3218  │ 5.00052  │    1    │   wheel   │   yes    │ 1    │   643.53307   │
│   6   │ csharp         │ 1        │ kinematics_poolq6            │  3165  │ 5.00078  │    1    │   wheel   │   yes    │ 1    │   632.90127   │
│   7   │ csharp         │ 1        │ kinematics_poold6            │  3093  │ 5.00050  │    1    │   wheel   │   yes    │ 1    │   618.53815   │
│   8   │ csharp         │ 1        │ kinematics_poolb6            │  2859  │ 5.00142  │    1    │   wheel   │   yes    │ 1    │   571.63765   │
│   9   │ csharp         │ 1        │ kinematics_bool2while        │  1880  │ 5.00361  │    1    │   base    │   yes    │ 8    │   375.72872   │
│  10   │ csharp         │ 1        │ kinematics_rawd2             │  1781  │ 5.00026  │    1    │   base    │   yes    │ 1    │   356.18148   │
│  11   │ csharp         │ 1        │ kinematics_bit2while         │  1636  │ 5.00208  │    1    │   base    │   yes    │ 1    │   327.06394   │
│  12   │ csharp         │ 1        │ kinematics_poold2            │  1633  │ 5.00063  │    1    │   base    │   yes    │ 1    │   326.55885   │
│  13   │ basic          │ 2        │ rbergen_vb                   │  1180  │ 5.00317  │    1    │   base    │   yes    │ 1    │   235.85024   │
└───────┴────────────────┴──────────┴──────────────────────────────┴────────┴──────────┴─────────┴───────────┴──────────┴──────┴───────────────┘
```

Not sure if it is worth it for so few succeeding, but maybe I can figure out why some of the others fail. However, I'll only spend more time on it if this is of any interest for the project. What say you?

**Comment by rbergen** [comment]
To be brutally honest, I don't think this would be of much interest. We already have a lower-specced machine in the set of runners. I also think that many solutions will refuse to run due to the memory limitation; 4GB on the host machine really is very tight for any Docker scenario and even more so on Windows.

What would be an interesting runner to add is a performant system with an AMD CPU. But of course, that's an entirely different cup of tea.

**Reply by rbergen** [reply]
Since my last comment, Dave has very kindly made a Ryzen 9 (5950X) based machine available that has been added to the runner pool.

**Reply by PEZ** [reply]
Thanks for your brutal honesty 😄 

Awesome with that extra runner. Now I must go check how it deals with the Clojure solutions.

**Reply by PEZ** [reply]
Probably a bit hard to achieve, but it would be quite interesting with a high end Apple Silicon machine. My MacBook with M1 Max processor, actually beats that Ryzen 9 5950X 16-Core for those extremely fast multithreaded Rust solutions.

```
bit-unrolled-hybrid            Passes: 391311, Threads: 6, Time: 5.0001101494, Passes / sec: 78260.48, Limit: 1000000, Counts: 78498, Valid: Pass

mike-barber_bit-unrolled-hybrid;391311;5.0001101494;6;algorithm=base,faithful=yes,bits=1
bit-extreme-hybrid             Passes: 398299, Threads: 6, Time: 5.0001840591, Passes / sec: 79656.87, Limit: 1000000, Counts: 78498, Valid: Pass

mike-barber_bit-extreme-hybrid;398299;5.0001840591;6;algorithm=base,faithful=yes,bits=1
```

Not sure why it only used 6 threads, when there are 8 cores available, but anyway. That's 13,276 passes/s/t, pretty crazy!

**Reply by mike-barber** [reply]
@PEZ that's awesome, thanks! That's pretty fast, although I'm more curious about the single-threaded results on the M1 personally ;) I don't have one to play around with, though. It sounds like it should be pretty much up there with the 5950X. 

We did have a problem with Dave's machine for some reason, but it looks to be fixed now and it's returning numbers like this in the latest run -- about 950k passes over 32 threads (16 physical cores)
```
mike-barber_bit-extreme-hybrid; 944723; 5.0009641647; 32
```

On my Ryzen 3900X machine, I get about 670k passes on 24 threads (12 physical cores)
```
mike-barber_bit-extreme-hybrid; 668035; 5.0008978844; 24
```

I'm not sure why it would be picking 6 threads automatically on the M1 Max. Might just be some updates required to some components to deal with the new processor. However, I do have command line switches to specify things like the number of threads if you're keen to play around with it a bit.

If you're running from the command line using `cargo` in the `PrimeRust/solution_1` folder, it's just 
```
cargo run --release -- --bits-extreme --threads 8
```

If you're using Docker, 
```
docker run --rm primerust_solution_1 --bits-extreme --threads 8
```

**Reply by PEZ** [reply]
I'm not sure how the OS allows access to the cores of the M1 on my machine... It is actually 10, but two of them seem to be treated special, always busy in a way that does not seem related to me compiling things or sieving primes and such. That's why I thought 8 threads would be the optimal.

Also not sure what Docker imposes in terms of overhead.  I get much better results using `cargo` directly.

Here are some sample runs:

Docker, default (1 and 6 threads):
```
docker run --rm rust-mike-barber --bits-extreme            

-------------------------------------------------------
Computing primes to 1000000 on 1 thread for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 78800, Threads: 1, Time: 5.0000061989, Passes / sec: 15759.98, Limit: 1000000, Counts: 78498, Valid: Pass


-------------------------------------------------------
Computing primes to 1000000 on 6 threads for 5 seconds.
-------------------------------------------------------

mike-barber_bit-extreme-hybrid;78800;5.0000061989;1;algorithm=base,faithful=yes,bits=1
mike-barber_bit-extreme-hybrid;395273;5.0002408028;6;algorithm=base,faithful=yes,bits=1
bit-extreme-hybrid             Passes: 395273, Threads: 6, Time: 5.0002408028, Passes / sec: 79050.79, Limit: 1000000, Counts: 78498, Valid: Pass
```

Docker, 8 threads:
```
docker run --rm rust-mike-barber --bits-extreme --threads 8 

-------------------------------------------------------
Computing primes to 1000000 on 8 threads for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 395730, Threads: 8, Time: 5.0011591911, Passes / sec: 79127.66, Limit: 1000000, Counts: 78498, Valid: Pass

mike-barber_bit-extreme-hybrid;395730;5.0011591911;8;algorithm=base,faithful=yes,bits=1
```

Docker, 10 threads:
```
docker run --rm rust-mike-barber --bits-extreme --threads 10

-------------------------------------------------------
Computing primes to 1000000 on 10 threads for 5 seconds.
-------------------------------------------------------

mike-barber_bit-extreme-hybrid;396544;5.0006089211;10;algorithm=base,faithful=yes,bits=1
bit-extreme-hybrid             Passes: 396544, Threads: 10, Time: 5.0006089211, Passes / sec: 79299.14, Limit: 1000000, Counts: 78498, Valid: Pass
```



Cargo default, (1 and 10 threads):
```
cargo run --release -- --bits-extreme            
    Finished release [optimized] target(s) in 0.05s
     Running `target/release/prime-sieve-rust --bits-extreme`

-------------------------------------------------------
Computing primes to 1000000 on 1 thread for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 86309, Threads: 1, Time: 5.0000238419, Passes / sec: 17261.72, Limit: 1000000, Counts: 78498, Valid: Pass
mike-barber_bit-extreme-hybrid;86309;5.0000238419;1;algorithm=base,faithful=yes,bits=1


-------------------------------------------------------
Computing primes to 1000000 on 10 threads for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 668878, Threads: 10, Time: 5.0001907349, Passes / sec: 133770.50, Limit: 1000000, Counts: 78498, Valid: Pass
mike-barber_bit-extreme-hybrid;668878;5.0001907349;10;algorithm=base,faithful=yes,bits=1
```

Cargo, 8 threads:
```
cargo run --release -- --bits-extreme --threads 8
    Finished release [optimized] target(s) in 0.08s
     Running `target/release/prime-sieve-rust --bits-extreme --threads 8`

-------------------------------------------------------
Computing primes to 1000000 on 8 threads for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 632547, Threads: 8, Time: 5.0001568794, Passes / sec: 126505.43, Limit: 1000000, Counts: 78498, Valid: Pass
mike-barber_bit-extreme-hybrid;632547;5.0001568794;8;algorithm=base,faithful=yes,bits=1
```

Cargo, 6 threads:
```
cargo run --release -- --bits-extreme --threads 6 
    Finished release [optimized] target(s) in 0.05s
     Running `target/release/prime-sieve-rust --bits-extreme --threads 6`

-------------------------------------------------------
Computing primes to 1000000 on 6 threads for 5 seconds.
-------------------------------------------------------

bit-extreme-hybrid             Passes: 493138, Threads: 6, Time: 5.0001149178, Passes / sec: 98625.34, Limit: 1000000, Counts: 78498, Valid: Pass
mike-barber_bit-extreme-hybrid;493138;5.0001149178;6;algorithm=base,faithful=yes,bits=1
```

The way that 8 vs 6 threads does not matter much using docker while it seems to just scale using cargo, I am thinking maybe with docker I am actually running it through Rosetta... Also, using cargo there is a slight difference between 8 and 10 threads, so maybe the OS is allowing the 2 "special” cores to be used some...

**Reply by mike-barber** [reply]
Nice one, thanks @PEZ 

Those are some pretty impressive numbers from the M1 Max. Cargo run, that is.

> I'm not sure how the OS allows access to the cores of the M1 on my machine... It is actually 10, but two of them seem to be treated special, always busy in a way that does not seem related to me compiling things or sieving primes and such. That's why I thought 8 threads would be the optimal.
> 
> Also not sure what Docker imposes in terms of overhead. I get much better results using cargo directly.

I'm also not sure -- it looks pretty costly on your machine. 

I can give a Windows (WSL2) perspective on the single-thread case (-t 1 --bits-extreme)
- 62819 - cargo
- 61048 - docker (~ 2.8% slower)

In the 24-thread case (-t 24 --bits-extreme)
- 666477 - cargo
- 648871 - docker (~ 2.6% slower)

Docker looks to have a much larger impact on your machine than this. I don't have a Mac to compare unfortunately. But if it is a Rosetta thing, perhaps it would be clear from checking what architecture the docker image was built for on your machine with `docker inspect`. On my machines, I get this back:
- `"Architecture": "amd64", "Os": "linux"` on the AMD machine
- `"Architecture": "arm64", "Variant": "v8", "Os": "linux"` on the raspberry pi

> The way that 8 vs 6 threads does not matter much using docker while it seems to just scale using cargo, I am thinking maybe with docker I am actually running it through Rosetta... Also, using cargo there is a slight difference between 8 and 10 threads, so maybe the OS is allowing the 2 "special” cores to be used some...

Agreed -- your numbers aren't _increasing_ with more cores in Docker. It looks maxed out at 6, as if the container engine only has that many cores assigned to it. This really could be the case -- as it would also explain why the program is auto-detecting 6 cores for the multithreaded run. Either way, it looks a bit broken in Docker on the Mac, since the single-threaded numbers are too slow too. 

The native performance, using cargo directly, is very impressive though. Let's just appreciate for a second that this is a _laptop_, and it's able to return numbers that on a similar level to Dave's Ryzen 5950X (single thread), and my Ryzen 3900X (full out all cores). These are desktop parts with cooling and well over 100W power limits.

| Case          | M1 Max (native) | Ryzen 3900X (mine, native) | Ryzen 5950X (dave, docker) |
|---------------|----------------:|---------------------------:|---------------------------:|
| 1 thread      | 86309           | 62819                      | 69480                      |
| all cores     | 668878          | 666477                     | 957974                     |

**Comment by PEZ** [comment]
One of the most amazing things with this laptop is that it never gets warm. Haven't ever heard the fans. Makes me wonder if there's a switch for allowing it to get warm. (Though, as I actually use it in my lap I am super happy with that it always stays cool. 😄)

Thanks for the tips about `docker inspect`. It gives me: `"Architecture": "arm64", "Variant": "v8", "Os": "linux"`. I still suspect we're seeing rosetta performance, even though I have installed docker from [here](https://docs.docker.com/desktop/mac/apple-silicon/) and should have native support...

**Reply by mike-barber** [reply]
> One of the most amazing things with this laptop is that it never gets warm. Haven't ever heard the fans. Makes me wonder if there's a switch for allowing it to get warm. (Though, as I actually use it in my lap I am super happy with that it always stays cool. 😄)

That's pretty awesome. In contrast, my work laptop (HP Z-book G5, Intel thing) pretty much runs fans up to maximum while I'm working and stays there all day. The cats love it. Me, less so: when it really get hot, it slows down significantly.

> Thanks for the tips about docker inspect. It gives me: "Architecture": "arm64", "Variant": "v8", "Os": "linux"

It does sound like it's building for the correct architecture at least (arm64), but I don't know what other layers of emulation docker has in place to run Linux images in OS X. At least when you're using cargo directly, you're getting an actual OS X executable. Anyway, it's a bit beyond me being not-an-Apple person ;)
